### DISCUSSION TITLE
How to update location

### DISCUSSION THREAD

**Discussion by ilarimikkonen** [HYPERTEGRITY-AG/OpenSim#2] [Show and tell]
Hi @WillFreitag cool tool!

I used it like this to update location:

```
#!/bin/bash
additional=1
limit=10
lng=16
lat=58
i=1; while [ $i -le $limit ]; do
  echo "The current lng is "$lng
  echo " lat is"$lat
./datasim.py -s http://orion:1026 --tenant magicbeans-y beanbag -ad dateObserved -an temperature,f,21.5 -al location,$lng,$lat -an batteryLevel,i,99 -an speed,i,12 -i -m 2
  lng=$(echo "scale=4; ( (${lng} * 100) + 1) / 100" | bc)
  lat=$(echo "scale=4; ( (${lat} * 100) + 1) / 100" | bc)
if [ $additional -gt 100 ]
  then
     lng=16
     additional=1
fi
#i=$((i+1))
additional=$((additional+1))
sleep 1;
done
```

floating match is kinda hard in shell scripting

**Comment by WillFreitag** [comment]
Hey @ilarimikkonen and many thanks!
Great idea to simulate a geo path!
I will think about a solution reading a kml-file (for example exported from google maps) and send that route time depended to an Orion Context Broker. 
Hope to give an example (bash) in the next days :-)

**Comment by WillFreitag** [comment]
Hey again @ilarimikkonen !
Got it :-)
So, first of all, I'm not using kml-files, but gpx-file-format.
With https://www.gpxgenerator.com you can easily create such a file describing a route with times per position (given an average speed for the route).
An example exported file looks like:
``` xml

<?xml version="1.0"?>
<gpx version="1.1" creator="gpxgenerator.com">
<wpt lat="51.726578514305174" lon="8.759472002022175">
    <ele>112.62</ele>
    <time>2021-03-30T09:53:58Z</time>
</wpt>
<wpt lat="51.72652534946479" lon="8.759300340645222">
    <ele>112.75</ele>
    <time>2021-03-30T09:53:59Z</time>
</wpt>
</gpx>
``` 
Using [xmlscarlet](http://xmlstar.sourceforge.net/doc/UG/xmlstarlet-ug.html) in a bash-script makes it easy to parse such a gpx-file.
Check the following bash-script printing out the content of a gpx-file with sleeping appropriate time between two waypoints:
``` bash
#!/bin/bash

# empty arrays for the values
lat=()
lon=()
time=()

# read latitudes
for l in $(xmlstarlet select -t -v '//wpt/@lat' sample.gpx)
do
   lat+=("$l")
done

# read longitudes
for l in $(xmlstarlet select -t -v '//wpt/@lon' sample.gpx)
do
   lon+=("$l")
done

# read times
for t in $(xmlstarlet select -t -v '//wpt/time' sample.gpx)
do
   time+=("$t")
done

# show what we've got
echo "Found "${#lat[@]}" latitudes."
echo "Found "${#lon[@]}" longitudes."
echo "Found "${#time[@]}" times."
echo

# TODO: check if the array sizes are identical

# loop though arrays
StartDate=$(date -u -d "${time[0]}" +"%s.%N")

for (( i=0; i<${#lat[@]}; i++ ))
do
  echo "Waypoint #$i:"
  # calculate sleep
  ThisDate=$(date -u -d "${time[$i]}" +"%s.%N")
  difference=$(awk '{print $1-$2}' <<< "$ThisDate $StartDate")
  echo "Sleeping for "$difference" second(s)"
  sleep $difference

  StartDate=$ThisDate
  echo "lat: ${lat[$i]}"
  echo "lon: ${lon[$i]}"
  echo
done
```
Having called OpenSim instead of echoing latitude and longitude will be an easy task.

I think it's worth a new "how to" for this use case...So feel free to play around with the script above or simply wait until the "howto.md" is written :-)

**Comment by WillFreitag** [comment]
You can quickstart with the two attached files (sample.gpx is a very short route and track.sh is the script from above).
Be aware to have `xmlstarlet` installed (e.g. `sudo apt-get install xmlstarlet` on Ubuntu).
Run the bash-script to see, how it continuously echos lat/long or exchange the echoing by (according to your exmple above):
``` bash
./datasim.py -s http://orion:1026 --tenant magicbeans-y beanbag -ad dateObserved -an temperature,f,21.5 -al location,${lat[$i]},${lon[$i]} -an batteryLevel,i,99 -an speed,i,12 -i -m 2
```


[example.zip](https://github.com/HYPERTEGRITY-AG/OpenSim/files/6229056/example.zip)

**Comment by WillFreitag** [comment]
Hey @ilarimikkonen .
Find a new "how to" [here](https://github.com/HYPERTEGRITY-AG/OpenSim/blob/main/HowTos/Send%20Route%20(gpx).md).
Many thanks for your inspiration!
